Add Copy. ignoreExistingContentInDestinationDir() (#18231)
gradle/gradle: subprojects/docs/src/docs/userguide/authoring-builds/working_with_files.adoc

You can also perform the same function in your own tasks with the link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:sync(org.gradle.api.Action)[Project.sync(org.gradle.api.Action)] method.

[[sec:copy_deploy]]
=== Deploying single files into application servers

When working with application servers, you can use a `Copy` task to deploy the application archive (e.g. a WAR file).
Since you are deploying a single file, the destination directory of the `Copy` is the whole deployment directory.
The deployment directory sometimes does contain unreadable files like named pipes, so Gradle may have problems doing up-to-date checks.

> In order to support this use-case, you can use link:{groovyDslPath}/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:doNotTrackDestinationDir()[Copy.doNotTrackDestinationDir()].

Reviewer: I think we should also mention the "too large unrelated outputs" use case here, too.

Author: Covered in the example for installing an executable.

.Using Copy to deploy a WAR file
====
include::sample[dir="snippets/files/deployWarWithCopy/groovy",files="build.gradle"]
include::sample[dir="snippets/files/deployWarWithCopy/kotlin",files="build.gradle.kts"]
====

> 

Reviewer: Is this too specific? Installing your runnable application code to `/usr/share` for example also qualifies here.

Author: That is one of the use-cases. Installing single files into directories with many other files is another use-case. `/usr/share/` doesn't seem to be the right example, though, since you'd install into a full sub-directory there, so you don't need untracked necessarily. I suppose `/usr/local/bin` would be the example? I can add this use-case as well.

Reviewer: Sorry, yes, I was thinking of `/usr/local/bin`.

Author: I added an example for installing an executable.

When working with application servers, you can use a `Copy` task to deploy the application archive (e.g. a WAR file).
Since you are deploying a single file, the destination directory of the `Copy` is the whole deployment directory.
The deployment directory sometimes does contain unreadable files like named pipes, so Gradle may have problems doing up-to-date checks.
In order to support this use-case, you can use link:{groovyDslPath}/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:ignoreExistingContentInDestinationDir()[Copy.ignoreExistingContentInDestinationDir()].

.Using Copy to deploy a WAR file
====
include::sample[dir="snippets/files/deployWarWithCopy/groovy",files="build.gradle"]
include::sample[dir="snippets/files/deployWarWithCopy/kotlin",files="build.gradle.kts"]
====

[[sec:install_executable]]
=== Installing executables

When you are building a standalone executable, you may want to install this file on your system, so it ends up in your path.

> You can use a `Copy` to install the executable into shared directories like `/usr/local/bin`.

Reviewer: ```suggestion
You can use a `Copy` task to install the executable into shared directories like `/usr/local/bin`.
```

The installation directory probably contains many other executables, some of which may even be unreadable by Gradle.

> To support the unreadable files in `Copy`'s destination directory and to avoid doing possible expensive up-to-date checks, you should use link:{groovyDslPath}/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:ignoreExistingContentInDestinationDir()[Copy.ignoreExistingContentInDestinationDir()].

Reviewer: ```suggestion
To support the unreadable files in the `Copy` task's destination directory and to avoid time consuming up-to-date checks, you can use link:{groovyDslPath}/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:ignoreExistingContentInDestinationDir()[Copy.ignoreExistingContentInDestinationDir()].
```
